idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,518 @@
+		// do a blocking write to the queue.
+		select {
+		case b.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}:
+			archive.LastSave = now
+			b.MemoryIdx.UpdateArchive(archive)
+		default:
+			statSaveSkipped.Inc()
+			log.Debugf("bigtable-idx: writeQueue is full, update of %s not saved this time", archive.MetricDefinition.Id)
+		}
+	}
+
+	return archive
+}
+
+func (b *BigtableIdx) rebuildIndex() {
+	log.Info("bigtable-idx: Rebuilding Memory Index from metricDefinitions in bigtable")
+	pre := time.Now()
+
+	var staleTs uint32
+	if b.cfg.MaxStale != 0 {
+		staleTs = uint32(time.Now().Add(b.cfg.MaxStale * -1).Unix())
+	}
+	num := 0
+	var defs []schema.MetricDefinition
+	for _, partition := range cluster.Manager.GetPartitions() {
+		defs = b.LoadPartition(partition, defs[:0], staleTs)
+		num += b.MemoryIdx.Load(defs)
+	}
+
+	log.Infof("bigtable-idx: Rebuilding Memory Index Complete. Imported %d. Took %s", num, time.Since(pre))
+}
+
+func (b *BigtableIdx) LoadPartition(partition int32, defs []schema.MetricDefinition, cutoff uint32) []schema.MetricDefinition {
+	ctx := context.Background()
+	rr := bigtable.PrefixRange(fmt.Sprintf("%d_", partition))
+	defsBySeries := make(map[string][]schema.MetricDefinition)
+	var marshalErr error
+	err := b.tbl.ReadRows(ctx, rr, func(r bigtable.Row) bool {
+		def := schema.MetricDefinition{}
+		marshalErr = RowToSchema(r, &def)
+		if marshalErr != nil {
+			return false
+		}
+		log.Debugf("bigtable-idx: found def %+v", def)
+		defsBySeries[def.Name] = append(defsBySeries[def.Name], def)
+		return true
+	}, bigtable.RowFilter(bigtable.FamilyFilter(COLUMN_FAMILY)))
+	if err != nil {
+		log.Fatalf("bigtable-idx: failed to load defs form Bigtable. %s", err)
+	}
+	if marshalErr != nil {
+		log.Fatalf("bigtable-idx: failed to marshal row to metricDef. %s", marshalErr)
+	}
+
+LOOP:
+	for series, defList := range defsBySeries {
+		for _, def := range defList {
+			if def.LastUpdate > int64(cutoff) {
+				// add all defs for this series.
+				defs = append(defs, defsBySeries[series]...)
+				continue LOOP
+			}
+		}
+		// all defs are stale
+		delete(defsBySeries, series)
+	}
+
+	return defs
+}
+
+func (b *BigtableIdx) processWriteQueue() {
+	timer := time.NewTimer(time.Second)
+	buffer := make([]writeReq, 0)
+
+	flush := func() {
+		rowKeys := make([]string, len(buffer))
+		mutations := make([]*bigtable.Mutation, len(buffer))
+		byteCount := 0
+		for i, req := range buffer {
+			statQueryInsertWaitDuration.Value(time.Since(req.recvTime))
+			key, cols := SchemaToRow(req.def)
+			rowKeys[i] = key
+			mut := bigtable.NewMutation()
+			for col, val := range cols {
+				mut.Set(COLUMN_FAMILY, col, bigtable.Now(), val)
+				byteCount += len(val)
+			}
+			mutations[i] = mut
+		}
+		statSaveBytesPerRequest.Value(byteCount)
+		pre := time.Now()
+		complete := false
+		attempts := 0
+		for !complete {
+			errs, err := b.tbl.ApplyBulk(context.Background(), rowKeys, mutations)
```

Comment:
does ApplyBulk not auto-retry? because i see we have our own retry logic, despite the context being used has no cancellation or deadline.